Changes required to build whole MB source

DoDoENT opened the pull request.

DoDoENT committed:
support for using GCC on Apple platform

DoDoENT committed:
[gcc_compatibles] use gnu++14 instead of gnu++1z

gnu++1z is not compatible with targets using Qt Framework

DoDoENT committed:
updated toolchain files with MB defaults

DoDoENT committed:
added support for enabling optimizations for host CPU

DoDoENT committed:
do not pass c++ flags to c compiler

DoDoENT committed:
use OS X 10.11 as minimum deployment target

DoDoENT committed:
these do not work as expected

DoDoENT committed:
if generator is Xcode, then set deployment target and let Xcode determine sysroot

DoDoENT committed:
avoid COMPILE_LANGUAGE generator expression on VS

DoDoENT committed:
use level 3 warnings by default in MSVC

DoDoENT committed:
perform Visual Studio version detection

lrotim committed:
fix gcc detection on linux

DoDoENT committed:
disable runtime checks in debug mode by default.

Each target should specifically turn this on if required.

DoDoENT committed:
added support toolchain for Android Studio's integrated toolchain

DoDoENT committed:
enabled gc-sections linker option for linux release build

DoDoENT committed:
disable RTTI by default

DoDoENT committed:
if APPLE is defined, undefine it

This is to prevent some confusions in CMake scripts while cross-compiling
to Android from Apple platform.

DoDoENT committed:
use gold linker on supported android platforms

DoDoENT committed:
set TNUN_ABI to ANDROID_ABI in android-studio toolchain

DoDoENT committed:
disable native CPU optimization when cross-compiling

psiha commented on toolchains/gcc_compatibles.cmake:
```diff
@@ -29,7 +29,7 @@ add_compile_options( -fstrict-aliasing -fstrict-enums -fvisibility=hidden -fvisi
 # + 'COMPILE_LANGUAGE' isn't supported by VS generators:
 # https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#logical-expressions
-add_compile_options( $<$<COMPILE_LANGUAGE:CXX>:-std=gnu++1z> )
+add_compile_options( $<$<COMPILE_LANGUAGE:CXX>:-std=gnu++14> )
```
why?

DoDoENT replied:
Qt 4.7 and PCH compatibility. When linking with targets created by Qt framework, they automatically publish compiler flag `-std=gnu++11` or `-std=gnu++14` to your code. OTOH, when assembling flags for PCH, this flag may still not be published (e.g. in cases target for which PCH is added depends on target which is not yet defined but which ultimately depends on Qt - PCH then ends up with `-std=gnu++1z`, whilst code ends up with `-std=gnu++14`. AFAIK, this issue can only be solved by CMake adding truly native support for PCH (without hacks we currently use).

psiha replied:
If we both agree to push for "/std:latest", can you find another way to hack around this? E.g. remove this option afterwards with remove_definitions or override it with another add_compile_options (or a directory or target specific version of it)?

DoDoENT replied:
I'll try to. However, the real fix for this issue would be for CMake guys to fix [issue 1260](https://gitlab.kitware.com/cmake/cmake/issues/1260).

psiha commented on build_options.cmake:
```diff
@@ -9,10 +9,12 @@
 # https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html
 # http://stackoverflow.com/questions/12802377/in-cmake-how-can-i-find-the-directory-of-an-included-file
-if ( WIN32 )
+if ( WIN32 ) # TODO: add detection of Windows Phone / Windows Universal platform
```
that would probably go with an explicit toolchain file (like android and ios)...

DoDoENT replied:
Not really - WP uses default MSVC toolchain.

For example, to configure code for WP 8.0 ARM, you should use:

```
cmake -G "Visual Studio 11 2012 ARM" -DCMAKE_SYSTEM_NAME=WindowsPhone -DCMAKE_SYSTEM_VERSION=8.0
```

For WP 8.1, similarly:

```
cmake -G "Visual Studio 12 2013 ARM" -DCMAKE_SYSTEM_NAME=WindowsPhone -DCMAKE_SYSTEM_VERSION=8.1
```

And for Windows Mobile 10.0 / Universal Windows App:

```
cmake -G "Visual Studio 14 2015 ARM" -DCMAKE_SYSTEM_NAME=WindowsStore -DCMAKE_SYSTEM_VERSION=10.0
```

Of course, specifiyng `Win64` instead of `ARM` will generate 64-bit VS project and not specifying anything will generate 32-bit (x86) VS project. On WindowsPhone, `x86` is usually interpreted as simulator build, while WindowsStore can be either `x86`, `ARM` or `Win64` (in Windows 8.0 and 8.1, WindowStore referred to Metro apps which used different MSVCRT than WindowsPhone - on Windows 10, WindowsStore system name refers to universal windows app).

psiha commented:
Thanks DoDo for the help and cooperation, unfortunately I cannot accept this PR as is since it contains MB specific changes...I will go through those in the commits and mark/comment them...after they are removed (really these should go into your specific build scripts) I'll merge the PR...

psiha commented on toolchains/osx.toolchain.cmake:
```diff
@@ -39,17 +39,40 @@ set( TNUN_os_suffix OSX )
 
 set( TNUN_cpu_archs default )
 
+set( TNUN_ABI x64 )
```
where do you need this?

DoDoENT replied:
In MB Source when we give names to tests, we add ABI suffix to be able to distinguish them in CI server report. In old build system this was PHOTOPAY_ARCHITECTURE variable which is now TNUN_ABI. On MacOS we always have x64 ABI.

Generally, this can be removed from build, but that would then require us to have parallel osx.toolchain.cmake file that would have to be loaded after this one.

psiha replied:
hm yes...universal build don't play well with this scheme...
so let's leave it like that for now...and later we can then add separate x86 and x86+x64 'unibuild' abis if needed...

psiha commented on toolchains/osx.toolchain.cmake:
```diff
@@ -39,17 +39,40 @@ set( TNUN_os_suffix OSX )
-#    set( XCODE_ATTRIBUTE_CFLAGS_i386   "-msse4.1 -march=core2 -mtune=core2"  )
-#    set( XCODE_ATTRIBUTE_CFLAGS_x86_64 "-msse4.1 -march=core2 -mtune=corei7" )
-#endif()
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
```
this whole block can only go in if it is enabled only for generators other than xcode (as it makes no sense for xcode)
- the MACOSX_CPP_LIBRARY\* variables can be named more generic like TNUN_CPP_LIB
- the clang+stdc++ and gcc+libc++ combinations and their permutations (if we really need them) can be made more generic (you can have the same situation on linux or even windows)...
- i would rather avoid the introduction of USING_CLANG and USING_GCC variables if possible

DoDoENT replied:
USING_CLANG and USING_GCC are for simplicity (to make following ifs easier to read) and can be removed.

I agree about TNUN_CPP_LIB variable which should generally be platform agnostic, but OTOH forcing to use specific STL with specific compiler requires flags specific to both compiler and platform (see line 71). This is something that should probably be defined in separate "toolchain add-on" files.

psiha replied:
since there is the builtin MSVC i'd rather go with CLANG and GCC then..

psiha commented on toolchains/msvc.toolchain.cmake:
```diff
@@ -26,20 +28,25 @@ set( TNUN_compiler_exceptions_off        -wd4577
   -D_SCL_SECURE_NO_WARNINGS
   -D_SBCS
   -D_WIN32_WINNT=0x0601 # Win7
+  -DNO_STRICT
```
this is known to break some software/libraries that depend on these...but ok let's leave it like that for now..

psiha commented on toolchains/msvc.toolchain.cmake:
```diff
@@ -26,20 +28,25 @@ set( TNUN_compiler_exceptions_off        -wd4577
 # w4373: '...': virtual function overrides '...', previous versions of the compiler did not override when parameters only differed by const/volatile qualifiers
 # w4324: 'structure was padded due to alignment specifier'
-add_compile_options( /std:c++latest -MP -Oi -W4 -Zc:threadSafeInit- -wd4324 -wd4373 -wd4100 -wd4456 -wd4067 -wd4995 -wd4635 -wd4800 -wd4101 -wd4756 -wd4068 -wd4056 -wd4996 -wd4503)
+add_compile_options( /std:c++latest -MP -Oi -Zc:threadSafeInit- )
```
why did you remove the default /W4? i consider this a feature of the build system to default to the 'compiler is your friend' attitude...
what is the reason for the introduction of TNUN_disabled_warnings?

DoDoENT replied:
We treat all warnings as errors and W4 introduces some very silly warnings which would then either need to be disabled with TNUN_disabled_warnings or fixed (in our source this would mean several weeks of work).

TNUN_disabled_warnings variable is added to contain silly warnings that should be disabled because they are both unimportant and difficult or impossible to fix. We treat other warnings as errors so adding those would make code uncompilable. AFAIK, MSVC does not have the ability to treat all warnings as errors, except some specifically defined (something like -Wno-error= on GCC/Clang). Even in that case, producing a lot of warnings, especially in third party libs, would produce a lot of noise in compile output, making it difficult for developers to find actual compile error they are facing (not everyone is using VS which separates warnings from errors in GUI).

psiha replied:
adding /w3 afterwards should acomplish what you wish?
and /wd should work the same as -Wno-error?

psiha commented on toolchains/linux/x86.abi.cmake:
```diff
@@ -0,0 +1,10 @@
+################################################################################
+
+add_compile_options( -m32 )
+link_libraries( -m64 )
```
m64?

psiha commented on toolchains/linux.toolchain.cmake:
```diff
@@ -0,0 +1,57 @@
+#
+################################################################################
+
+# MACOSX_RPATH
```
leftovers? :)

DoDoENT replied:
yep

psiha commented on toolchains/linux.toolchain.cmake:
```diff
@@ -0,0 +1,57 @@
+    include( "${CMAKE_CURRENT_LIST_DIR}/gcc_compatibles.cmake" )
+endif()
+
+option( TNUN_STATIC_LIBC "Link all binaries statically with libc and libc++" false )
```
preferably this would be a more generic option (e.g. also often used on windows)...but ok we can add this in a separate change..

psiha commented on toolchains/ios.toolchain.cmake:
```diff
@@ -20,6 +20,8 @@
 
 include( "${CMAKE_CURRENT_LIST_DIR}/apple.cmake" )
 
+add_compile_options( $<$<COMPILE_LANGUAGE:CXX>:-x objective-c++> )
```
this is a hack...which is sometimes necessary but then one should do it for a specific file in his own build scripts...this cannot go in as a default....that's what the .mm file extension is for...

psiha commented on toolchains/gcc_compatibles.cmake:
```diff
@@ -22,9 +23,16 @@ set( TNUN_compiler_exceptions_off      -fno-exceptions                       )
 set( TNUN_compiler_report_optimization -ftree-vectorizer-verbose=6           )
-set( TNUN_compiler_release_flags       -DNDEBUG -fomit-frame-pointer -ffunction-sections -fdata-sections -fmerge-all-constants -fno-stack-protector )
+set( TNUN_compiler_release_flags       -DNDEBUG -fomit-frame-pointer -ffunction-sections -fmerge-all-constants -fno-stack-protector )
+if( ${CMAKE_SYSTEM_NAME} MATCHES "Linux" )
```
why not put this into the linux toolchain?

DoDoENT replied:
The original idea was to also have this flag on android, but since it was added in android toolchain, it can be removed from here.

psiha commented on toolchains/gcc_compatibles.cmake:
```diff
@@ -22,9 +23,16 @@ set( TNUN_compiler_exceptions_off      -fno-exceptions                       )
+set( TNUN_native_optimization          -march=native -mtune=native           )
 
-add_compile_options( -fstrict-aliasing -fstrict-enums -fvisibility=hidden -fvisibility-inlines-hidden -fno-threadsafe-statics -Wall -Wstrict-aliasing -Wno-multichar -Wno-unknown-pragmas -Wno-unused-local-typedefs )
+add_compile_options( -fstrict-aliasing -fstrict-enums -fvisibility=hidden -fvisibility-inlines-hidden -fPIC )
```
where did  -fno-threadsafe-statics go?

psiha commented on toolchains/gcc_compatibles.cmake:
```diff
@@ -22,9 +23,16 @@ set( TNUN_compiler_exceptions_off      -fno-exceptions                       )
+endif()
+set( TNUN_default_warnings             -Wall -Wextra -Wstrict-aliasing       )
+set( TNUN_warnings_as_errors           -Werror                               )
+set( TNUN_disabled_warnings            -Wno-unknown-pragmas -Wno-unused-parameter -Wno-ignored-qualifiers -Wno-sign-conversion -Wno-clobbered -Wno-format-zero-length -Wno-error=deprecated-declarations     )
```
where do you have problems with Wclobbered?
Wunused-parameter is not a trivial warning and i don't want it disabled by default...but again why the introduction of this TNUN_disabled_warnings variable?

DoDoENT replied:
-Wunused-parameter is very annoying when overriding virtual functions and not using all parameters.

-Wno-clobbered is probably leftover from some legacy code or support for compiling some third party libraries we use in PhotoPay (kontocheck would be my first guess) - I'll have to recheck whether everyhing works without it.

psiha replied:
no go @ -Wunused-parameter, that's really the basics in tidy code ... simply don't give the parameter a name (or put its name in a /**/ comment block...standard practice)...+ plus you can add that option locally..

psiha commented on toolchains/gcc_compatibles.cmake:
```diff
@@ -9,9 +9,10 @@
 # http://stackoverflow.com/questions/31355692/cmake-support-for-gccs-link-time-optimization-lto
 set( CMAKE_INTERPROCEDURAL_OPTIMIZATION true )
-set( CMAKE_VISIBILITY_INLINES_HIDDEN    true )
+#set( CMAKE_VISIBILITY_INLINES_HIDDEN    true )
```
?

DoDoENT replied:
Protobuf has problems when this is enabled - macOS linker throws lots of warnings.

psiha replied:
hidden visibility should really be standard practice these days...can't you disable it just for protobuf?

psiha commented on toolchains/gcc.cmake:
```diff
@@ -0,0 +1,20 @@
+
+list( APPEND TNUN_disabled_warnings -Wno-error=cpp -Wno-deprecated-declarations -Wno-unused-result )
+
+if( ${CMAKE_SYSTEM_NAME} MATCHES "Linux" )
```
why just linux?

psiha replied:
?

psiha commented on toolchains/clang.cmake:
```diff
@@ -12,9 +12,16 @@ include( "${CMAKE_CURRENT_LIST_DIR}/gcc_compatibles.cmake" )
 # http://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
 # http://clang.llvm.org/docs/UsersManual.html#controlling-code-generation
-set( TNUN_compiler_runtime_sanity_checks -fsanitize=undefined -fsanitize=integer )
+set( TNUN_compiler_runtime_sanity_checks -fsanitize=undefined -fsanitize=integer -fsanitize=address -fno-omit-frame-pointer )
```
why do you need -fno-omit-frame-pointer here?

psiha replied:
?

DoDoENT replied:
To get nicer stack trace in sanitization report. See [documentation](http://clang.llvm.org/docs/AddressSanitizer.html#usage)

psiha commented on toolchains/clang.cmake:
```diff
@@ -12,9 +12,16 @@ include( "${CMAKE_CURRENT_LIST_DIR}/gcc_compatibles.cmake" )
+set( TNUN_compiler_runtime_sanity_checks -fsanitize=undefined -fsanitize=integer -fsanitize=address -fno-omit-frame-pointer )
 
-add_compile_options( -Wheader-guard )
+if( "${CMAKE_CXX_COMPILER}" MATCHES ".*ccache" )
```
can you add an implementation note explaining this block?

DoDoENT replied:
When clang is used behind ccache, it throws a lot of "unused-argument" warnings. I'm not sure why that happens (it also happens on clang for Android, both from ndk-build and cmake android build). Having this much warnings for each and every source file is very annoying.

psiha replied:
ok but can you add that as an implementation note inside the clang.cmake file?

psiha commented on build_options.cmake:
```diff
@@ -52,8 +54,27 @@ function( TNUN_add_compile_options configuration )
+    add_compile_options( ${TNUN_compiler_rtti_on} )
+else()
+    add_compile_options( ${TNUN_compiler_rtti_off} )
+endif()
```
not to happy about this block: in ide's/multiconfig generators you'll often want rtti in debug but not in release and also, although that is less common, the same for exceptions..

DoDoENT committed:
fix typos

DoDoENT committed:
apparently --icf=all does not work for mips targets

DoDoENT committed:
define DEBUG in gcc-compatible compilers in debug mode

psiha commented on build_options.cmake:
```diff
@@ -14,7 +14,9 @@ if ( WIN32 ) # TODO: add detection of Windows Phone / Windows Universal platform
 elseif( ${CMAKE_SYSTEM_NAME} MATCHES "Linux" )
-    include ( "${CMAKE_CURRENT_LIST_DIR}/toolchains/linux.toolchain.cmake" )
+    include( "${CMAKE_CURRENT_LIST_DIR}/toolchains/linux.toolchain.cmake" )
+elseif( ANDROID_TOOLCHAIN ) # TNUN android.toolchain.cmake does not define this variable, while native Android Studio toolchain does
```
aren't you essentially including the android toolchin file twice here?

DoDoENT replied:
Essentially yes. Android Studio first forces its own android toolchain which does not have as optimal flags as TNUN android cmake. Since you cannot force AS to use non-default toolchain, this includes a special "android-studio toolchain" which improves flags imposed by AS default toolchain.

psiha commented on build_options.cmake:
```diff
@@ -63,17 +65,39 @@ if ( ${TNUN_DEBUG_SYMBOLS_IN_RELEASE} )
-    add_compile_options( ${TNUN_compiler_exceptions_on} )
+option( TNUN_ALLOW_RTTI "Allow Runtime Type Information in C++ code" false )
+
+if( ${CMAKE_GENERATOR} MATCHES "Visual Studio" )
```
huh...why did this explode?

psiha commented on toolchains/android-studio.toolchain.cmake:
```diff
@@ -0,0 +1,66 @@
+#
+# T:N.U.N. Android Studio CMake tool chain file. This is to be used together with
+# native android CMake toolchain bundled with Android studio 2.2 and newer.
+# For uses without Android Studio, please use android.toolchain.cmake file.
```
can't these two toolchains be merged into one?

psiha commented on toolchains/android.toolchain.cmake:
```diff
@@ -228,6 +228,7 @@ set( CMAKE_OBJDUMP      "${ANDROID_TOOLCHAIN_ROOT}/bin/${_CMAKE_TOOLCHAIN_PREFIX
 set( CMAKE_RANLIB       "${ANDROID_TOOLCHAIN_ROOT}/bin/${_CMAKE_TOOLCHAIN_PREFIX}ranlib${TOOL_OS_SUFFIX}"  )
 
 include( "${CMAKE_CURRENT_LIST_DIR}/clang.cmake" )
+unset( TNUN_native_optimization ) # This makes no sense when cross-compiling.
```
why not simply not-give the option when crosscompiling in the first place?

psiha commented on toolchains/apple.cmake:
```diff
@@ -41,9 +41,9 @@ set( TNUN_ABI   default    )
-set( CMAKE_XCODE_ATTRIBUTE_PROJECT_TEMP_DIR "${PROJECT_BINARY_DIR}" )
-set( CMAKE_XCODE_ATTRIBUTE_SYMROOT          "${PROJECT_BINARY_DIR}/lib" )
-set( CMAKE_XCODE_ATTRIBUTE_SYMROOT_RELEASE  "${PROJECT_BINARY_DIR}/lib" )
+# set( CMAKE_XCODE_ATTRIBUTE_OBJROOT          "${PROJECT_BINARY_DIR}" )
```
you have give a reason for commenting things out?

psiha commented on toolchains/msvc.toolchain.cmake:
```diff
@@ -30,7 +30,7 @@ set( TNUN_compiler_optimize_for_speed    -Ox -Ot -Qpar
 set( TNUN_disabled_warnings              -wd4324 -wd4373 -wd4100 -wd4456 -wd4067 -wd4995 -wd4635 -wd4800 -wd4101 -wd4756 -wd4068 -wd4056 -wd4996 -wd4503 )
 set( TNUN_warnings_as_errors             -WX )
-set( TNUN_default_warnings               -W4 )
+set( TNUN_default_warnings               -W3 )
```
/W4 has to be the default...

psiha commented on toolchains/msvc.toolchain.cmake:
```diff
@@ -64,6 +64,21 @@ endif()
 
+# Detect Visual Studio Version
+if( ${CMAKE_GENERATOR} MATCHES "Visual Studio" )
+    if(${CMAKE_CXX_COMPILER_VERSION} VERSION_LESS 20 AND ${CMAKE_CXX_COMPILER_VERSION} VERSION_GREATER 19)
```
why do you need this?

psiha commented on toolchains/osx.toolchain.cmake:
```diff
@@ -17,10 +17,11 @@ include( "${CMAKE_CURRENT_LIST_DIR}/apple.cmake" )
-set( CMAKE_OSX_SYSROOT                 "macosx"                      ) #"Latest Mac OS X"
-set( CMAKE_OSX_DEPLOYMENT_TARGET       "10.7"                        )
+if( ${CMAKE_GENERATOR} MATCHES "Xcode" )
+    set( CMAKE_OSX_DEPLOYMENT_TARGET "10.11"                       )
```
no can do...other people/projects still have to support older osx versions...
10.7 has libcpp by default and that is enough as a base minimum...
and why the condition on xcode?

DoDoENT replied:
I agree on 10.7 - I've put 10.11 because MB 3rd party libs are built with 10.11 min target and causes linker warnings. I agree this should be overriden in our source.

Setting `CMAKE_OSX_DEPLOYMENT_TARGET` on makefile/ninja generator causes problems with PCH.

DoDoENT committed:
as noted by psiha, this should be in linux toolchain, not here

DoDoENT committed:
added clang implementation note for -Qunused-arguments

DoDoENT committed:
added linker runtime sanitization flags

without them, linux linker fails with:
    undefined reference to `__ubsan_handle_sub_overflow'
    undefined reference to `__asan_report_load8'
    undefined reference to `__asan_init'
    ...
    ... and much more

DoDoENT committed:
also add thread and leak sanitizer and add correct linker flags

DoDoENT committed:
unset TNUN_linker_runtime_sanity_checks on OS X as this is not supported there

DoDoENT committed:
when allowing debug symbols in release, make sure linker flags are also passed

DoDoENT committed:
use TNUN_add_link_options after defining it

DoDoENT committed:
disable sanitization on android for now

I will investigate what is wrong later.

DoDoENT committed:
support detection of AppleClang vs normal Clang

This enables support for address and undefined sanitization on OS X, if
Clang from homebrew is used.

DoDoENT committed:
disable leak sanitizer for now, as it crashes with MB binaries

I am currently not sure whether the crash is due to a memory leak in our
code or bug in leak sanitizer itself. The same binary works correctly under
valgrind. Even weirder - the ASAN:DEADLYSIGNAL is obtained in benign
pthread_create call.

This needs to be investigated further at a later time.

DoDoENT committed:
Setup flags for leak and address sanitization.

Apparently errors reported earlier were due to some other reason.
This configuration should now work on both OS X and Linux (and possibly
on clang on windows).

DoDoENT committed:
set CMAKE_CROSSCOMPILING to true for android toolchain

DoDoENT committed:
disable clang leak sanitizer for now

It's very very slow (two orders of magnitude slower than valgrind).

DoDoENT committed:
experimental support for ARMv7 hardfloat ABI

DoDoENT committed:
added support for precompiled headers

DoDoENT committed:
removed variables TNUN_ALLOW_EXCEPTIONS, TNUN_ALLOW_RTTI and TNUN_NATIVE_OPTIMIZATION

Those compiler flags should be used by 3rd party projects as they wish.
We should not enforce global usage of those flags across all projects.

DoDoENT committed:
removed variable TNUN_disabled_warnings

3rd parties should take care for disabling warnings in their projects.

DoDoENT committed:
restored original behaviour in gcc_compatibles and osx and msvc toolchains

DoDoENT committed:
support for choosing any (Clang,GCC)-(stdc++,libc++) combination on Linux and OSX

DoDoENT committed:
fixes

- added support for multiple targets in single binary dir
- extract flags from target properties, not only from directory property

DoDoENT committed:
disable add_compile_options( -std=gnu++1z ) because it conflicts with CMAKE_CXX_STANDARD

DoDoENT committed:
add STL flags only for C++ files

DoDoENT committed:
added note for PCH

Commit:
disable /std:c++latest flag in VS2015u3, as it causes compilation problems

psiha commented:
Google is your friend ;P
https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler
#define _HAS_AUTO_PTR_ETC 1
(I actually though this was already in the build somewhere...)

I'd like to document this build system as 'by default' promoting latest C++ practices and language advancements, Boost, high warning levels etc...
(and yes for now it requires MSVC15u3/latest VS, latest Xcode)...

DoDoENT committed:
Revert "disable /std:c++latest flag in VS2015u3, as it causes compilation problems"

This reverts commit ae0466d6f032ad07bf080e9744c4b589bb0587d5.

DoDoENT committed:
[pch] add support for generator expressions in include_dirs and compile_definitions

However, single generator expression containing multiple elements is still
not supported because it breaks the foreach loop.

DoDoENT committed:
support for armv7s in universal build script and support for multiple targets

multiple targets are supported by preventing deleting of result folder
each time script is invoked

DoDoENT committed:
set compiler ID in ios toolchain

DoDoENT committed:
do not add any compile flags globally

This gives opportunity for 3rd parties to remove some default compiler
flags before calling add_compile_options.

DoDoENT committed:
fix linking for armeabi-v7a-hard

DoDoENT committed:
fix detection of iOS 10 SDK
